types.go

```diff
@@ -108,20 +108,16 @@ type FunctionInfo struct {
 
 // functionInfo contains information about which interfaces the objective
 // function F implements and the actual methods of F that have been
-// successfully type switched.
+// successfully type switched. TODO(jds) change this comment
 type functionInfo struct {
```

Comment:
Sure, I don't object to having FunctionInfo, I just needed a gentle reminder why we have it. So we keep it, but I would rename it to ProblemInfo and the fields from IsGradient to HasGradient etc. "Is" came from Function _being_ Gradient, etc interface.